[Theodore] Duke Increments

theodoreleebrant opened the pull request:
In fulfillment of CS2103 requirement.

Commit:
Add Gradle support

theodoreleebrant committed:
Level 1 commit
Add greet, echo, and exit function

theodoreleebrant committed:
Level 1 commit
Add greet, echo, and exit function

theodoreleebrant committed:
Fix delimiter bug
Change next() to nextLine() for scanner as the only delimiter that should be accepted is a newline character.

theodoreleebrant committed:
Edit delimiter bug and merge head
change scanner method to nextLine since the only delimiter which should be accepted is newline character.

theodoreleebrant committed:
Finish Level 2
Add the ability to store whatever text entered by the user and display them back to the user when requested.

theodoreleebrant committed:
Add list function

theodoreleebrant committed:
Finish Level 3
Add the ability to mark tasks as done.

theodoreleebrant committed:
Complete Level 4 tasks.
Add three types of tasks:
    ToDos: tasks without any date/time attached to it e.g., visit new theme park
    Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
    Events: tasks that start at a specific time and ends at a specific time e.g., team project meeting on 2/10/2019 2-4pm
Change formatting of output

theodoreleebrant committed:
Add parser

theodoreleebrant committed:
fix bug for todo

theodoreleebrant committed:
Implement error handling
Use exception to handle errors.

theodoreleebrant committed:
indentation bugfix

theodoreleebrant committed:
Level 6 submission

theodoreleebrant committed:
Add case-insensitive handling

theodoreleebrant committed:
Refactor code to adhere to OOP

theodoreleebrant committed:
Add ability to save the tasks in the hard disk automatically when task list closes

theodoreleebrant committed:
Resolve conflict

theodoreleebrant committed:
Resolve conflict (try 2)

theodoreleebrant committed:
Revert "Resolve conflict (try 2)"

This reverts commit a25744c98233580a728aba356f93f918694975b5.

theodoreleebrant committed:
Add support for date and time

theodoreleebrant committed:
Merge remote-tracking branch 'origin/branch-Level-7'

theodoreleebrant committed:
Merge branch 'branch-Level-8'

theodoreleebrant committed:
Implement searching

theodoreleebrant committed:
Add JUnit tests

theodoreleebrant committed:
Edit braces to comply with coding standard

theodoreleebrant committed:
Add Javadocs header comment

theodoreleebrant committed:
Put files in package

theodoreleebrant committed:
remove todo

theodoreleebrant committed:
Add more javadocs

theodoreleebrant committed:
Cleanup 'find' code

theodoreleebrant committed:
Merge branch 'branch-A-CodingStandard'

theodoreleebrant committed:
Merge remote-tracking branch 'origin/branch-Level-9'

theodoreleebrant committed:
Merge remote-tracking branch 'origin/branch-A-JavaDoc'

theodoreleebrant committed:
Add a Command interface and corresponding classes
Refactor code to include the Command interface and corresponding classes
Storage is to be re-implemented in a future date

theodoreleebrant committed:
Fix coding standard after refactoring

raysonkoh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.ArrayList;
+
+public class TaskList {
+    public ArrayList<Task> thingsToDo;
```
How about renaming `thingsToDo` to `tasks` to convey the same meaning?

raysonkoh commented on src/main/java/duke/ListCommand.java:
```diff
@@ -0,0 +1,11 @@
+
+public class ListCommand implements Command {
+    public ListCommand() {}
+    @Override
```
Would it be better to have a linebreak above this `Override` statement?

raysonkoh commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,21 @@
+import java.time.format.DateTimeParseException;
+
+public class Deadline extends Task {
+    LocalDateTime timeBy;
```
Would it be better to rename `timeBy` to `deadline`?

raysonkoh commented on src/main/java/duke/EventCommand.java:
```diff
@@ -0,0 +1,19 @@
+package duke;
+
+public class EventCommand implements Command {
+    Task toBeAdded;
```
Could `toBeAdded` be a `private` `final` variable?

raysonkoh commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.Date;
+
+public class Event extends Task {
+    public LocalDateTime timeAt;
```
Could `timeAt` be a `private` `final` variable?

cx0810 replied:
Perhaps the variable name could be atDateTime instead

raysonkoh commented on src/main/java/duke/DeadlineCommand.java:
```diff
@@ -0,0 +1,19 @@
+package duke;
+
+public class DeadlineCommand implements Command {
+    Task toBeAdded;
```
Would renaming `toBeAdded` to `newDeadlineTask` be better?

raysonkoh reviewed (commented):
Looks good! :+1:

Apart from minor nits in naming variables and access modifiers, there are not much coding standard violations. You might want to consider moving some related classes to a subpackage (eg classes related to `Command`). Also, I think it might be helpful to write some tests for the core classes like `Storage` and `TaskList`.

I liked how you used regex in your `Parser` class!

cx0810 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    public static Task parseDeadline(String echo) throws DukeException {
+        if (echo.matches("(?i)^deadline\\s+\\S+.*\\s+\\/by\\s+\\S+.*")) {
```
Perhaps can simplify this or use another way to express it? Looks very complicated.

theodoreleebrant replied:
Unfortunately no, the regex used forces the following:

`(?i)`: case insensitive matching
`^`: at the start of the command,
`deadline`: match the word `deadline` exactly
`\\s+`: followed by at least one space character
`\\S+`: followed by at least one non-space character (i.e. the description)
`.*`: followed by anything (to support for multi-word descriptions)
`\\s+`: followed by at least one space character
`\\/`: followed by the character `/`
`by`: followed by `by`
`\\s+`: followed by at least one space character
`\\S+`: followed by at least one non-space character (i.e. there is a time)
`.*`: followed by anything (to support for multi-word descriptions)

cx0810 reviewed (commented):
LGTM.

theodoreleebrant committed:
Increase granularity in Task Type

theodoreleebrant committed:
Merge branch 'add-gradle-support'

theodoreleebrant committed:
Configure gradle
Delete checkstyle configuration
Update Duke package configuration

theodoreleebrant committed:
Configure JavaFX for GUI

theodoreleebrant committed:
Integrate FXML in JavaFX

theodoreleebrant committed:
Integrate logic into GUI

theodoreleebrant committed:
Revert "Integrate logic into GUI"

This reverts commit 01ba2509a172c275963fb2b5c824016db1e8d36b.

theodoreleebrant committed:
Improve error handling.

theodoreleebrant committed:
Merge remote-tracking branch 'origin/branch-Level-10'

theodoreleebrant committed:
Change readme file

theodoreleebrant committed:
Merge pull request #1 from theodoreleebrant/readme

Change readme file

theodoreleebrant committed:
Fix coding style violation

theodoreleebrant committed:
Merge pull request #2 from theodoreleebrant/branch-A-CodeQuality

Fix coding style violation

theodoreleebrant committed:
Add assertion

theodoreleebrant committed:
Merge pull request #3 from theodoreleebrant/branch-A-Assertions

Add assertion

theodoreleebrant committed:
Add help message

theodoreleebrant committed:
Create gradle.yml for CI

theodoreleebrant committed:
Create gradle.yml

theodoreleebrant committed:
Remove wrong location of gradle CI

theodoreleebrant committed:
Update README.md

theodoreleebrant committed:
Add Ui.png

theodoreleebrant committed:
Update README.md

theodoreleebrant committed:
Set theme jekyll-theme-hacker

theodoreleebrant committed:
Update README.md

theodoreleebrant committed:
Update README.md

theodoreleebrant committed:
Optimize for build

theodoreleebrant committed:
Add comment

theodoreleebrant committed:
Refactor code

Add javadoc comments
Organize file into packages
Optimization for javafx

theodoreleebrant committed:
Improve GUI

Add icon
Add padding to text
Add title to application

theodoreleebrant committed:
Remove useless imports

theodoreleebrant committed:
Improve UX: Allow users to copy text field

javafx label does not allow selection of text.
This is removed and replaced by a TextArea.
Additionally, the UI is improved with new colors.

theodoreleebrant committed:
Make time optional for Event & Deadline

theodoreleebrant committed:
Add new checkstyle file

theodoreleebrant committed:
Delete test file

theodoreleebrant committed:
Make date parser smart

The date parser now autocorrects wrong date, to a certain extent.
For example, 30 February is auto-corrected to 29 February.

theodoreleebrant committed:
Refactor storage

theodoreleebrant committed:
Enable resizing window

theodoreleebrant committed:
Bugfix: add space for delete and done command

theodoreleebrant committed:
Invoke storage for every command

theodoreleebrant committed:
Fix bug in storage

theodoreleebrant committed:
Improve text returned for find

theodoreleebrant committed:
Replace Ui.png

theodoreleebrant committed:
Update README with new UI

theodoreleebrant committed:
Update README.md

theodoreleebrant committed:
Update README.md

theodoreleebrant committed:
Update README.md

theodoreleebrant committed:
Update README.md

theodoreleebrant committed:
Set theme jekyll-theme-minimal
